### PR TITLE
[NAE-1497] Frontend actions

### PR DESCRIPTION
**Pull request by renczesstefan**
# Description

Implemented frontend actions and interprocess communication regarding ``change`` and ``make`` actions.

Implements [NAE-1497]

## Dependencies

No new dependencies where introduced.

### Third party dependencies

No new dependencies where introduced.

### Blocking Pull requests

There are no dependencies on other PR.

## How Has Been This Tested?

This was tested manually and with unit tests.

### Test Configuration

| Name                | Tested on |
|---------------------| --------- |
| OS                  |   macOS Ventura 13.2     |
| Runtime             |  Java 11        |
| Dependency Manager  |  Maven 3.8.4         |
| Framework version   |  Spring Boot 2.6.2        |
| Run parameters      |           |
| Other configuration |           |

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes have been checked, personally or remotely, with @minop @RudeBoyxX 
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have resolved all conflicts with the target branch of the PR
- [x] I have updated and synced my code with the target branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes:
    - [x] Lint test
    - [x] Unit tests
    - [x] Integration tests
- [ ] I have checked my contribution with code analysis tools:
    - [ ] [SonarCloud](https://sonarcloud.io/project/overview?id=netgrif_application-engine)
    - [ ] [Snyk](https://app.snyk.io/org/netgrif/project/ea82b3b8-658d-41f5-89a7-76cd600da01f)
- [ ] I have made corresponding changes to the documentation:
    - [ ] Developer documentation
    - [ ] User Guides
    - [ ] Migration Guides

[NAE-1497]: https://netgrif.atlassian.net/browse/NAE-1497?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy [hunk 1]
```diff
@@ -458,59 +474,89 @@ class ActionDelegate {
         }]
     }
 
-    protected SetDataEventOutcome createSetDataEventOutcome() {
-        return new SetDataEventOutcome(this.useCase, this.task.orElse(null))
+    def makeInterProcess(List<String> fieldIds, Closure behavior) {
```
Comment (Retoocs) [inline src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:477]: docstring should be added along with examples

**Review by Retoocs — CHANGES_REQUESTED** [review]

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy (hunk 1 above)
Comment (Retoocs) [inline src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:477]: Is it necessary to have new function name? Cannot be overloaded existing `make` function?

Comment (renczesstefan) [reply src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:477]: Added.

Comment (renczesstefan) [reply src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:477]: ``make`` has an input argument of type ``List`` and ``Closure`` and this function has the same types. If they have the same name, Java/Groovy is not able to decide, which one to use.

**Review by Retoocs — APPROVED** [review]

File: docs/actions/frontend-actions.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+Frontend.<action_name>(Object args...)
+```
+
+Example:
+```
+<event type="set">
+    <id>data_set_event</id>
+    <actions phase="post">
+        <action id="action_0">
+            Frontend.redirec("login")
```
Comment (mazarijuraj) [inline docs/actions/frontend-actions.md:18]: Missing `t` in `redirect`

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy [hunk 3]
```diff
@@ -458,59 +474,104 @@ class ActionDelegate {
...
+     *     def taskIds = [taskRef.value[0]] as List
+     *     makeInterProcess ["field_Id"], editable on taskIds when {true}
+     *
+     * </pre>
+     * This code will change the behavior of fields <i>text</i> and <i>anotherText</i> to <i>visible</i> on given transition when field's <i>condition</i> value is equal to <i>true</i>.
+     *
+     * @param list of field IDs which behaviour will be changed
+     * @param behavior one of initial, visible, editable, required, optional, hidden, forbidden
+     */
+    def makeInTask(List<String> fieldIds, Closure behavior) {
```
Comment (mazarijuraj) [inline src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:492]: 1. Why is this a separate function and not part of original `make`?
2. Name of the function in javadoc is `makeInterProcess` not  `makeInTask`
3. Message of thrown exception is also misleading with `make` as the called method

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/Frontend.groovy [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    List<EventOutcome> outcomes
+
+    Frontend(Case useCase, Optional<Task> task, List<EventOutcome> outcomes) {
+        this.useCase = useCase
+        this.task = task
+        this.outcomes = outcomes
+    }
+
+    def methodMissing(String name, Object args) {
+        ChangedField changedField = new ChangedField(FRONTEND_ACTIONS_KEY)
```
Comment (mazarijuraj) [inline src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/Frontend.groovy:26]: Frontend actions should have own attribute in outcome and not be obscured in `chagedField`

**Review by mazarijuraj — CHANGES_REQUESTED** [review]

File: docs/actions/frontend-actions.md (hunk 2 above)
Comment (renczesstefan) [reply docs/actions/frontend-actions.md:18]: Corrected

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy (hunk 3 above)
Comment (renczesstefan) [reply src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:492]: `make` has the same argument types than this function, so they cannot have common name. Other things will be correected.

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/Frontend.groovy (hunk 4 above)
Comment (renczesstefan) [reply src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/Frontend.groovy:26]: Own attribute in `outcome` object? Then should it be only in `SetDataEventOutcome` or in root `EventOutcome`?

Comment (renczesstefan) [reply src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/Frontend.groovy:26]: Okay, I think I understand what you mean. I will refactor it.

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy [hunk 5]
```diff
@@ -2457,4 +2438,14 @@ class ActionDelegate {
         Case taskCase = workflowService.findOne(task.caseId)
         return taskCase.getPetriNet().getDataSet().get(fieldId)
     }
+
+    def test2() {
```
Comment (mazarijuraj) [inline src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:2442]: Please remove test function

**Review by mazarijuraj — CHANGES_REQUESTED** [review]

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy [hunk 6]
```diff
@@ -421,15 +428,29 @@ class ActionDelegate {
...
                                 fields.each { field ->
                                     if (trans.dataSet.containsKey(field.stringId)) {
                                         behaviorClosureResult = behavior(field, trans)
-                                        saveFieldBehavior(field, trans, (behavior == initial) ? behaviorClosureResult as Set : null)
+                                        saveFieldBehavior(field as Field, trans, (behavior == initial) ? behaviorClosureResult as Set : null)
+                                    }
+                                }
+                            } else if (trans instanceof Task) {
+                                fields.forEach { field ->
+                                    saveFieldBehaviorWithTask(field as Field<?>, trans, behavior, behaviorClosureResult)
```
Comment (mazarijuraj) [inline src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:444]: Missing javadoc example for calling `make` with field Ids from another case. Please add example for both if branches (trans is Task, trans is String)

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy (hunk 5 above)
Comment (mazarijuraj) [reply src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:2442]: Please remove test function

**Review by mazarijuraj — CHANGES_REQUESTED** [review]
Please add javadoc example of new `make` functionality.
`test2()` function is still present in ActionDelegate

**Review by mazarijuraj — APPROVED** [review]

**Pull request merged**